src/process/shell.cr

```diff
@@ -158,4 +176,66 @@ class Process
 
     tokens
   end
+
+  # Splits the given *line* into individual command-line arguments according to
+  # Microsoft's standard C runtime, unquoting tokens if necessary.
+  #
+  # Raises `ArgumentError` if a quotation mark is unclosed. Leading spaces in
+  # *line* are ignored. Otherwise, this method is equivalent to
+  # [`CommandLineToArgvW`](https://docs.microsoft.com/en-gb/windows/win32/api/shellapi/nf-shellapi-commandlinetoargvw)
+  # for some unspecified program name.
+  #
+  # NOTE: This does **not** take strings that are passed to the CMD shell or
+  # used in a batch script.
+  #
+  # ```
+  # Process.parse_arguments_windows(%q[foo"bar \\\"hello\\" Fizz\Buzz]) # => ["foobar \\\"hello\\", "Fizz\\Buzz"]
+  # ```
+  def self.parse_arguments_windows(line : String) : Array(String)
+    tokens = [] of String
+    reader = Char::Reader.new(line)
+    quote = false
+
+    while true
+      # skip whitespace
+      while reader.current_char.in?(' ', '\t')
+        reader.next_char
+      end
+      break unless reader.has_next?
+
+      token = String.build do |str|
+        while true
+          backslash_count = 0
+          while reader.current_char == '\\'
+            backslash_count += 1
+            reader.next_char
+          end
+
+          if reader.current_char == '"'
+            (backslash_count // 2).times { str << '\\' }
+            if backslash_count.odd?
+              str << '"'
+            else
+              quote = !quote
+            end
+            reader.next_char
+          elsif backslash_count > 0
+            backslash_count.times { str << '\\' }
+          elsif reader.has_next?
+            char = reader.current_char
+            break if char.in?(' ', '\t') && !quote
+            str << char
+            reader.next_char
+          else
+            break
+          end
```

Comment:
Suggestion for restructuring the flow a bit.
```suggestion
          else
            if backslash_count > 0
              backslash_count.times { str << '\\' }
            end
            if reader.has_next?
              char = reader.current_char
              break if char.in?(' ', '\t') && !quote
              str << char
              reader.next_char
            else
              break
            end
          end
```
IMO this makes it a bit clearer.
And it should also be a little bit more efficient. There's no need to loop directly from the `backslash_count > 0` branch, we can just continue with reading the next char (it can't be a backslash or double quote at this point).